text/0000-cargo-libstd-awareness.md

```diff
@@ -0,0 +1,236 @@
+We have an obligation to not break packages depending only on stable interfaces, so the implicit dependencies will include both `std` and `core`:
+```toml
+[dependencies]
+core = { version = "^1.0", stdlib = true }
+std  = { version = "^1.0", stdlib = true }
+
+[dev-dependencies]
+core = { version = "^1.0", stdlib = true }
+std  = { version = "^1.0", stdlib = true }
+test = { version = "^1.0", stdlib = true }
+
+[build-dependencies]
+core = { version = "^1.0", stdlib = true }
+std  = { version = "^1.0", stdlib = true }
+```
+The version requirement for `core` of `^1.0` may seem odd because core was not stable at the time, but anything else would either break newer packages using core, or prevent older packages from working on versions of Rust predating core's stabilization.
+Remember that rustc only complains if an unstable crate is actually imported, so the Cargo dependency on its own is harmless.
+
+`test` is a similar scenario.
+While importing it explicitly remains unstable, it's currently injected and thus needs to be built.
+Other dependencies of `std` besides core we don't need to worry about, because they are only transitive dependencies through `std`, not direct dependencies.
+
+Now, not all crates depend on these crates, so there must be a way to opt out.
+For this, we introduce a new `implicit-dependencies` key.
+It is defined by default as:
+```toml
+implicit-dependencies = ["primary", "build", "dev"]
+```
+This indicates each of `dependencies`, `build-dependencies`, and `dev-dependencies` maps (respectively) is augmented with implicit elements.
+A manual definition may be that or almost any subset, in which case only the included dependency maps are augmented.
+
+Finally, if an (explicit) dependency conflicts with one of the implicit defaults, that category of implicit dependency will be skipped.
+For example, if a crate explicit depends on `std` as a build-dependency, neither `std` nor any other implicit build dependency will be injected.
+This final rule means must packages won't need to use the implicit-dependencies key, because either implicit dependencies will be used, or a `core` or `std` explicit dependency will be present.
+An exception is `core` itself, which must of course not depend on `core` or `std` implicitly or explicitly.
+
+## Compiler command-line interface
+
+As will be explained, with these changes Cargo will always use `--extern` to pass dependencies, so rustc's sysroot lookup can and should be bypassed.
+This will prevent users using the sysroot from depending on more crates than their cargo file declares (implicitly or explicitly)..
+A `--no-resolve-sysroot` flag will be added for this purpose.
+Compilers that don't have sysroot binaries should accept this flag and ignore it.
+Once [Rust PR #35021](https://github.com/rust-lang/rust/pull/35021/files) lands in some form, `compiler-rt` will be a Cargoized dependency so rustc won't use the sysroot for linking either.
```

Comment:
Now that I think of it, there is still libunwind and jemalloc to worry about. Plus the general question of how transitive dependencies are dealt with I've referenced in a few places. I'll fix that once that is dealt with.